### ISSUE TITLE
Track recording doesn't record in the background on Android 8

### ISSUE THREAD

**Issue by PaulStets** [osmandapp/Osmand#5587]
Multiple Android 8 users complain that the track isn't recorded in the background starting from 3.0.2.

**Comment by sonora** [comment]
Let me cross-link to https://github.com/osmandapp/Osmand/issues/5255 and https://github.com/osmandapp/Osmand/commit/0158fa496eaa111bf0b8a9b57374460f2cc87524 for debugging this.

I would have thought that Victor's commit https://github.com/osmandapp/Osmand/commit/0158fa496eaa111bf0b8a9b57374460f2cc87524 as per Android documentation has fixed it under Android 8, but have no test device for that yet.

For Android 5 to 7, see my https://github.com/osmandapp/Osmand/issues/5255#issuecomment-388279138., I have devices where even disabling all Power Management results only in one point recorded every 5 minutes.

Input needed from users would be the recording interval setting used (>30 sec or less), and their (sometimes manufacturer-specific) Android Power management settings.

**Comment by vshcherb** [comment]
I have Android 8. Are we talking about long switch-off GPX recording (more than 15 minutes delay)?
For default settings I tested works fine.

**Comment by matthewblain** [comment]
I am seeing something like this; hopefully this info is helpful.

Device: Moto X4. OS: Android 8.1.0 OPW28.2 OsmAnd+ 3.0.4. 
Recording a track using the 5 second default.

* Recording usually works fine. 
* The notifications usually work, get updated, buttons to take action (e.g. pause) work on both lock screens. (I don't know what it's called, but the one which shows up on proxmity sensor and the one which shows up when you hit the power button).
* Recording works for a while..
* Taking a photo stops it. Sometimes. Specifically, I start recording, put phone in pocket, forget about it. Then take out the phone, twist (moto gesture to turn on camera), take photo, lock screen, put back in pocket.
-> Now the notifications are gone. It's like the 'background' app was stopped.
-> If I unlock the phone and go to OsmAnd+, it will resume recording and the notification from that point. Everything is as before--that is, on one of my tracks it recorded for a while until I took another photo. At least I think it stopped then.

One time I also got a notification that OsmAnd was using lots of battery and there was some setting to throttle it, but the default (which I left) was not to do so.

This happened to me a few times, including once today. But of course when I tried to make it happen again later today it failed to do so.

If there's some sort of useful log I can pull from my phone, let me know.

**Comment by sonora** [comment]
Thanks for explaining! It would be interesting to know if the behavior is connected to the device's Power saving settings or not, i.e. if you can reproduce ever while Power saving is turned off.

**Comment by matthewblain** [comment]
What power saving settings do you mean? The only two I can find are:
 "Optimizing Battery Use", which is on the default setting for this app.
 "Battery Saving" global setting, which is set to never (the default).

**Comment by sonora** [comment]
Sounds like you already answered it, but maybe you can test with disabilng power saving for OsmAnd on app level just to be sure?

**Comment by WebFreak001** [comment]
I don't know if I am experiencing the same issue, but I enabled the "ask to enable tracking in notification" option but it only appears once I start OsmAnd. Isn't it supposed to show at system startup or so instead? I would love to have it enabled & starting automatically on every startup/unlock if possible (I have OsmAnd on my Android Car radio and using the HTTP feature to track my position with it)

**Comment by vshcherb** [comment]
Since I installed Android 8, OsmAnd doesnot work properly anymore. It stops recording trip when I turn off screen.
My device is a Huawei P100 pro.

**Comment by bperel** [comment]
Reproduced the issue as well. Phone model is Huawei Mate 10 Pro (BLA-L29), I disabled both power optimization and Battery saving, but it looks like 5 minutes or so after I turn on the screen the GPS tracking stops being recorded. Please ask if I can provide more information.

**Comment by matthewblain** [comment]
FYI I changed the setting for the app about "Optimizing Battery Use" and... I think it happens less often?
The same bug (recording pauses when photos taken) happens with Trailforks and Strava too. Strava seems to have some sort of additional wake-up thing which causes it to start again on its own after (some time) and display a message that it 'crashed'.

**Comment by st-anton** [comment]
Huawei Mate9, EMUI 9.1.0 Android 9
I tested the tracking of OSMAND and Geo-Tracker.org in parallel. That is, both apps tracks the same gps locations at the same time. The power settings of Adroid are identical: Manual settings and continue to execute in background.

While Geo-Tracker tracked the complete way from start to finishing the tracking, OSMAND stopped the tracking at an undefined point and did not continue to track. I did some actions during the mapping like checking the mailbox, take photos and so on. Geo-Tracker simply continued.

In my eyes, the problem is not, that GPS stops to continue, but OSMAND stops to continue in background. Maybe Ilya Bogdanovich has a hint...

**Comment by sonora** [comment]
All Android apps published in the last year or so need to use a so called "Foreground Service" (which will be visible to the user in the Android Notification bar) in order to be allowed requesting frequent location updates even while other apps are in the foreground or the device screen is off. There"s not many degrees of freedom here, from what I know, all apps are forced to do the same.

What you could check under Android's app settings is if perhaps Battery optimization has been disabled for Geo Tracker, but not for OsmAnd? If yes  maybe you want to exempt OsmAnd also and repeat the experiment?

**Comment by st-anton** [comment]
> What you could check under Android's app settings is if perhaps Battery optimization has been disabled for Geo Tracker, but not for OsmAnd? If yes maybe you want to exempt OsmAnd also and repeat the experiment?
As I mentioned, the battery option is identical for both apps: Set optimization to manual (not handled automatically by Android) and allow to execute in background. Geo-Tracker as well as OsmAnd request to do the same settings in the battery (power) settings.

**Comment by st-anton** [comment]
Yes, I'm using latest versions and 'prevent standalone recording' is disabled.

I found three conditions when this happens:
1. randomly when I start another app e.g. the camera. This does not happen all the time, I start another app, but it happens noticable. I'm not sure whether it appears only if the other app was not already in the background, i.e. starting new.
2. when gps signal is lost, e.g. when entering a building
3. randomly / by chance, at least I see not relation to 1 or 2

Most of the time, the gps location 'stucks', even if the app itself reacts (pan and zoom). But the gps location is not found and it takes some time until the connection is found again. Note: The other app continuously tracks the correct location in the background! The gps signal ist NOT lost! I think this effect is seen in the past as 'the gps receiver was shut down and need to relocate after switching on again'. But it is not like this, osmand lost the 'software connection' to the gps service. (Sorry, I don't know the software architecture of Linux).

Notifications: When tracking is active and running, the 'snake' in the notification bar is flickering and the notification message is active. When recoding is 'stuck', the snake is still present, but no more flickering. The notification message is still present, but will not be updated.

**Comment by sonora** [comment]
Could you use Menu > Configure screen to place the GPS info widget on the map screen? What does it show during those periods? 0/0, or some other numbers?

PS: Case 2 above is clear: During periods without GPS reception no points can be recorded, and the GPS info should show 0/0, or perhaps 0/x. But points should be recorded immediately afterwards again.

PS2: Some apps may use low accuracy location methods like cell phone towers or WiFi networks to derermine location, OsmAnd (and most other navigation services) does not. Please also test "Google location service" vs. "Android API" in OsmAnd's settings, if not already done so. The first heavily depends on your version of Google Play Services, the second not.

Maybe open a new issue with all relevant information, if the problem persists. It seems to be system specific, I cannot reproduce, nor am I aware of other sinular active issues.
